### PR TITLE
Use device monitor to discover media sources

### PR DESCRIPTION
**Pull request by spira-inhabitant**
Use device monitor to list sources audio and video.
Use gst_device_create_element() to create elements.

### CODE REVIEW

File: local/owr_local_media_source.h [hunk 1]
```diff
@@ -54,6 +55,8 @@ typedef struct _OwrLocalMediaSourcePrivate OwrLocalMediaSourcePrivate;
 struct _OwrLocalMediaSource {
     OwrMediaSource parent_instance;
 
+    GstDevice *device;
```
Comment (philn) [inline local/owr_local_media_source.h]: This should be stored in the priv structure

File: local/owr_audio_renderer.c [hunk 2]
```diff
@@ -123,19 +123,27 @@ OwrAudioRenderer *owr_audio_renderer_new(void)
 static void
 setup_sink_for_aec(GstElement *sink)
 {
-#if defined(__linux__) && !defined(__ANDROID__)
+#if defined(HAVE_PULSE)
     /* pulsesink */
     GstStructure *s;
 
+    if (g_strcmp0(GST_OBJECT_NAME(gst_element_get_factory(src)), "pulsesink") != 0)
```
Comment (philn) [inline local/owr_audio_renderer.c]: There is no such src variable in this context.

File: local/owr_device_list.c [hunk 3]
```diff
@@ -297,103 +145,63 @@ static gboolean enumerate_audio_source_devices(GClosure *callback)
...
+    caps = gst_caps_new_any();
 
-    if (ioctl(fd, VIDIOC_QUERYCAP, &vcap) < 0) {
-        g_warning("v4l: failed to query %s", filename);
+    if (type == OWR_MEDIA_TYPE_AUDIO)
+        klass = "Source/Audio";
+    else if (type == OWR_MEDIA_TYPE_VIDEO)
+        klass = "Source/Video";
+    else
+        g_assert_not_reached();
```
Comment (philn) [inline local/owr_device_list.c:162]: Please initialize caps and monitor after this.

File: local/owr_device_list.c [hunk 4]
```diff
@@ -297,103 +145,63 @@ static gboolean enumerate_audio_source_devices(GClosure *callback)
...
-            return NULL;
-
-        source = _owr_local_media_source_new_cached(index, device_name,
-            OWR_MEDIA_TYPE_VIDEO, OWR_SOURCE_TYPE_CAPTURE);
-
-        g_debug("v4l: filename match: %s", device_name);
-
-        g_free(device_name);
-    }
+    _owr_utils_call_closure_with_list(callback, sources);
```
Comment (philn) [inline local/owr_device_list.c:188]: No need to reverse the sources list before calling this?

File: local/owr_local_media_source.c [hunk 5]
```diff
@@ -694,27 +708,33 @@ static GstElement *owr_local_media_source_request_source(OwrMediaSource *media_s
...
+                    CREATE_ELEMENT(source, AUDIO_SRC, "audio-source");
+
 #if !defined(__APPLE__) || !TARGET_IPHONE_SIMULATOR
 /*
     Default values for buffer-time and latency-time on android are 200ms and 20ms.
     The minimum latency-time that can be used on Android is 20ms, and using
     a 40ms buffer-time with a 20ms latency-time causes crackling audio.
     So let's just stick with the defaults.
 */
-#if !defined(__ANDROID__)
```
Comment (philn) [inline local/owr_local_media_source.c:705]: Can these changes be avoided?

File: local/owr_local_media_source.c [hunk 6]
```diff
@@ -744,17 +764,22 @@ static GstElement *owr_local_media_source_request_source(OwrMediaSource *media_s
...
             switch (source_type) {
             case OWR_SOURCE_TYPE_CAPTURE:
-                CREATE_ELEMENT(source, VIDEO_SRC, "video-source");
-                if (priv->device_index > -1) {
+                if (local_source->device) {
+                    /* source is provided by device monitor */
+                    source = gst_device_create_element(local_source->device, "video-source");
+                } else {
+                    CREATE_ELEMENT(source, VIDEO_SRC, "video-source");
+                    if (priv->device_index > -1) {
```
Comment (philn) [inline local/owr_local_media_source.c]: Why isn't this done for the device case too?

File: local/owr_local_media_source.h [hunk 7]
```diff
@@ -37,6 +37,7 @@
 #include "owr_types.h"
 
 #include <glib-object.h>
+#include <gst/gst.h>
```
Comment (philn) [inline local/owr_local_media_source.h]: Nit: GstDevice could be forward-declared instead.

**Review by philn — CHANGES_REQUESTED** [review]

File: local/owr_local_media_source.h (hunk 1 above)
Comment (spira-inhabitant) [reply local/owr_local_media_source.h]: Ok. Will be updated in next pull request.

File: local/owr_local_media_source.h (hunk 7 above)
Comment (spira-inhabitant) [reply local/owr_local_media_source.h]: N/A with device declared in priv structure.

File: local/owr_local_media_source.c (hunk 6 above)
Comment (spira-inhabitant) [reply local/owr_local_media_source.c]: You're right. Will be updated in next pull request.

File: local/owr_local_media_source.c (hunk 5 above)
Comment (spira-inhabitant) [reply local/owr_local_media_source.c:705]: Will be simplified in next pull request.

File: local/owr_device_list.c (hunk 4 above)
Comment (spira-inhabitant) [reply local/owr_device_list.c:188]: I think not. The list from device monitor is not supposed to be sorted.

File: local/owr_device_list.c (hunk 3 above)
Comment (spira-inhabitant) [reply local/owr_device_list.c:162]: Ok. Will be updated in next pull request.

File: local/owr_audio_renderer.c (hunk 2 above)
Comment (spira-inhabitant) [reply local/owr_audio_renderer.c]: Ok. Will be updated in next pull request.

**Comment by spira-inhabitant** [comment]
Pull request updated.

**Review by philn — APPROVED** [review]

**Pull request merged**